async/amqp/amqp.go

```diff
@@ -54,17 +54,47 @@ func (m *message) Nack() error {
 	return err
 }
 
-// Factory of a AMQP consumer
+// Exchange describes the Name and Kind of a RabbitMQ exchange
+type Exchange struct {
+	Name string
+	Kind string
+}
+
+func (ex Exchange) validate() error {
+	if ex.Name == "" {
+		return errors.New("RabbitMQ exchange Name is required")
+	}
+
+	if ex.Kind == "" {
+		return errors.New("RabbitMQ exchange type is required")
+	}
+
+	if ex.Kind != amqp.ExchangeDirect &&
+		ex.Kind != amqp.ExchangeFanout &&
+		ex.Kind != amqp.ExchangeTopic &&
+		ex.Kind != amqp.ExchangeHeaders {
+
+		return errors.New(fmt.Sprintf("RabbitMQ exchange type is invalid, one of [%s, %s, %s, %s] is required",
+			amqp.ExchangeDirect,
+			amqp.ExchangeFanout,
+			amqp.ExchangeTopic,
+			amqp.ExchangeHeaders))
+	}
+
+	return nil
+}
+
+// Factory of an AMQP consumer
 type Factory struct {
 	url      string
 	queue    string
-	exchange string
+	exchange Exchange
+	bindings []string
 	oo       []OptionFunc
 }
 
 // New constructor.
-func New(url, queue, exchange string, oo ...OptionFunc) (*Factory, error) {
-
+func New(url, queue string, exchange Exchange, bindings []string, oo ...OptionFunc) (*Factory, error) {
```

Comment:
we should add bindings as a option.
if no option is set then set binding to empty string.

Reply:
We already have a variadic argument so I used an empty string as binding if an empty slice is provided, like you mentioned

Reply:
if you check out the option.go in the package you see how to create a option the is the last variadic argument. this is called "functional configuration"...

Reply:
Made the bindings an OptionFunc